pkg/crds/apis/serverless/v1alpha1/realtimeapi_types.go

```diff
@@ -0,0 +1,283 @@
+	// can be an absolute number, e.g. 5, or a percentage of desired replicas, e.g. 10% (default: 25%)
+	// (set to 0 to disable rolling updates)
+	MaxSurge intstr.IntOrString `json:"max_surge"`
+
+	// +kubebuilder:validation:Optional
+	// maximum number of replicas that can be unavailable during an update; can be an absolute number,
+	// e.g. 5, or a percentage of desired replicas, e.g. 10%
+	MaxUnavailable intstr.IntOrString `json:"max_unavailable"`
+}
+
+type NetworkingSpec struct {
+	// +kubebuilder:validation:Optional
+	// Endpoint for the API
+	Endpoint string `json:"endpoint,omitempty"`
+}
+
+// RealtimeAPIStatus defines the observed state of RealtimeAPI
+type RealtimeAPIStatus struct {
+	// +kubebuilder:validation:Optional
+	// Number of ready pods
+	Ready int32 `json:"ready"`
+
+	// +kubebuilder:validation:Optional
+	// Number of requested pods
+	Requested int32 `json:"requested"`
+
+	// +kubebuilder:validation:Optional
+	// Number of pods with the last requested spec
+	UpToDate int32 `json:"up_to_date"`
+
+	// +kubebuilder:validation:Optional
+	// URL of the deployed API
+	Endpoint string `json:"endpoint,omitempty"`
+}
+
+//+kubebuilder:object:root=true
+//+kubebuilder:subresource:status
+//+kubebuilder:printcolumn:JSONPath=".status.ready",name="Ready",type="integer"
+//+kubebuilder:printcolumn:JSONPath=".status.requested",name="Requested",type="integer"
+//+kubebuilder:printcolumn:JSONPath=".status.up_to_date",name="Up-To-Date",type="integer"
+//+kubebuilder:printcolumn:JSONPath=".status.endpoint",name="Endpoint",type="string"
+
+// RealtimeAPI is the Schema for the realtimeapis API
+type RealtimeAPI struct {
+	kmeta.TypeMeta   `json:",inline"`
+	kmeta.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   RealtimeAPISpec   `json:"spec,omitempty"`
+	Status RealtimeAPIStatus `json:"status,omitempty"`
+}
+
+// GetOrCreateAPIIDs retrieves API ids from annotations or creates them if they don't exist
+func (api RealtimeAPI) GetOrCreateAPIIDs() (deploymentID, podID, specID, apiID string) {
+	deploymentID = api.Annotations["cortex.dev/deployment-id"]
+	if deploymentID == "" {
+		deploymentID = k8s.RandomName()[:10]
+	}
+
+	var buf bytes.Buffer
+
+	buf.WriteString(api.Name)
+	buf.WriteString(userconfig.RealtimeAPIKind.String())
+	buf.WriteString(s.Obj(api.Spec.Pod))
+	podID = hash.Bytes(buf.Bytes())
+
+	buf.Reset()
+	buf.WriteString(podID)
+	buf.WriteString(s.Obj(api.Spec.Networking))
+	buf.WriteString(s.Obj(api.Spec.Autoscaling))
+	buf.WriteString(s.Obj(api.Spec.NodeGroups))
+	buf.WriteString(s.Obj(api.Spec.UpdateStrategy))
+	specID = hash.Bytes(buf.Bytes())[:32]
+
+	apiID = api.Annotations["cortex.dev/api-id"]
+	if apiID == "" ||
+		api.Annotations["cortex.dev/deployment-id"] != deploymentID ||
+		api.Annotations["cortex.dev/spec-id"] != specID {
+		apiID = fmt.Sprintf("%s-%s-%s", spec.MonotonicallyDecreasingID(), deploymentID, specID)
+	}
+	return
```

Comment:
Do not do empty returns. We don't use that pattern anywhere else. It is better to have the explicit return, always